File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int find(vector<int>& inorder, int target, int start, int end) {
        for(int i = start; i <= end; i++) {
            if(inorder[i] == target)
                return i;
        }
        return -1;
    }

    TreeNode* buildTreeHelper(vector<int>& inorder, vector<int>& postorder, int inStart, int inEnd, int& postIndex) {
        if(inStart > inEnd)
            return NULL;

        TreeNode* root = new TreeNode(postorder[postIndex]);
        int pos = find(inorder, postorder[postIndex], inStart, inEnd);
        postIndex--;
  
        root->right = buildTreeHelper(inorder, postorder, pos + 1, inEnd, postIndex);
        root->left = buildTreeHelper(inorder, postorder, inStart, pos - 1, postIndex);
        
        return root;
    }

    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        int n = postorder.size();
        int postIndex = n - 1;  
        return buildTreeHelper(inorder, postorder, 0, n - 1, postIndex);
    }
};

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {

        return build(inorder, postorder, 0, inorder.length - 1, 0, postorder.length - 1);
        
    }
    public TreeNode build(int[] inorder, int[] postorder, int inStart, int inEnd, int postStart, int postEnd){

    if(inStart > inEnd || postStart > postEnd){
        return null;
    }

    int rootValue = postorder[postEnd];

    TreeNode root = new TreeNode(rootValue);

    int index = inStart;

    while(inorder[index] != rootValue){
        index++;
    }

    int leftsize = index - inStart;

    root.left = build(inorder ,postorder , inStart, index - 1, postStart, postStart + leftsize - 1);

    root.right = build(inorder, postorder,index + 1,inEnd,postStart + leftsize,postEnd - 1);

    return root;
}

}